network/src/lib.rs

```diff
@@ -3,61 +3,130 @@ pub mod propagation_network;
 
 use async_trait::async_trait;
 use serde::{Deserialize, Serialize};
-use simperby_common::crypto::*;
-use std::net::SocketAddrV4;
-use tokio::sync::broadcast;
+use simperby_common::{crypto::*, BlockHeight};
+use std::{net::SocketAddrV4, sync::Arc};
+use thiserror::Error;
+use tokio::sync::{mpsc, RwLock};
 
-pub type BroadcastToken = u64;
+#[derive(Error, Debug, PartialEq, Eq)]
+pub enum Error {
+    #[error("unknown error: {0}")]
+    Unknown(String),
+}
+
+pub struct Peer {
+    pub public_key: PublicKey,
+    pub address: SocketAddrV4,
+    /// An arbitrary string that the peer has set for itself.
+    /// This is usually used for indicating ports for the other services
+    /// that ths peer is running (e.g., Git, RPC, Message, ...)
+    pub message: String,
+}
+
+#[async_trait]
+pub trait PeerDiscovery {
+    /// Remains online on the network indefinitely,
+    /// responding to discovery requests from other nodes,
+    /// updating `known_peers`.
+    async fn serve(
+        network_config: &NetworkConfig,
+        known_peers: Arc<RwLock<Peer>>,
```

Comment:
Vec<Peer>?

Reply:
Nope. It will be constantly updated and should be accessible from the caller side.
In general, the `Arc` will be cloned and delivered to `P2PNetwork::serve()`.